actors/builtin/market/market_test.go

```diff
@@ -1 +1,347 @@
+			testCases := []struct {
+				delta int64
+				total int64
+			}{
+				{10, 10},
+				{20, 30},
+				{40, 70},
+			}
+
+			// Test adding provider funds from both worker and owner address
+			for _, callerAddr := range []address.Address{owner, worker} {
+				rt, actor := setup()
+
+				for _, tc := range testCases {
+					rt.SetCaller(callerAddr, builtin.AccountActorCodeID)
+					rt.SetReceived(abi.NewTokenAmount(tc.delta))
+					actor.expectProviderControlAddress(rt, providerIdAddr, owner, worker)
+
+					rt.Call(actor.AddBalance, &provider)
+
+					rt.Verify()
+
+					rt.GetState(&st)
+					assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, providerIdAddr))
+				}
+			}
+		})
+
+		t.Run("fails unless called by an account actor", func(t *testing.T) {
+			rt, actor := setup()
+
+			rt.SetReceived(abi.NewTokenAmount(10))
+			actor.expectProviderControlAddress(rt, providerIdAddr, owner, worker)
+
+			rt.SetCaller(provider, builtin.StorageMinerActorCodeID)
+			rt.ExpectAbort(exitcode.ErrForbidden, func() {
+				rt.Call(actor.AddBalance, &provider)
+			})
+
+			rt.Verify()
+		})
+
+		t.Run("adds to non-provider escrow funds", func(t *testing.T) {
+			testCases := []struct {
+				delta int64
+				total int64
+			}{
+				{10, 10},
+				{20, 30},
+				{40, 70},
+			}
+
+			// Test adding non-provider funds from both worker and client addresses
+			for _, callerAddr := range []address.Address{client, worker} {
+				rt, actor := setup()
+
+				for _, tc := range testCases {
+					rt.SetCaller(callerAddr, builtin.AccountActorCodeID)
+					rt.SetReceived(abi.NewTokenAmount(tc.delta))
+					rt.ExpectValidateCallerType(builtin.CallerTypesSignable...)
+
+					rt.Call(actor.AddBalance, &callerAddr)
+
+					rt.Verify()
+
+					rt.GetState(&st)
+					assert.Equal(t, abi.NewTokenAmount(tc.total), st.GetEscrowBalance(rt, callerAddr))
+				}
+			}
+		})
+	})
+
+	t.Run("WithdrawBalance", func(t *testing.T) {
+		t.Run("fails with a negative withdraw amount", func(t *testing.T) {
+			rt, actor := setup()
+
+			params := market.WithdrawBalanceParams{
+				ProviderOrClientAddress: provider,
+				Amount:                  abi.NewTokenAmount(-1),
+			}
+
+			rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+				rt.Call(actor.WithdrawBalance, &params)
+			})
+
+			rt.Verify()
+		})
+
+		t.Run("withdraws from provider escrow funds and sends to owner", func(t *testing.T) {
+			rt, actor := setup()
+
+			actor.addProviderFunds(rt, providerIdAddr, owner, worker, abi.NewTokenAmount(20))
+
+			rt.GetState(&st)
+			assert.Equal(t, abi.NewTokenAmount(20), st.GetEscrowBalance(rt, providerIdAddr))
+
+			// worker calls WithdrawBalance, balance is transferred to owner
+			rt.SetCaller(worker, builtin.AccountActorCodeID)
+			actor.expectProviderControlAddress(rt, providerIdAddr, owner, worker)
+
+			withdrawAmount := abi.NewTokenAmount(1)
+			rt.ExpectSend(owner, builtin.MethodSend, nil, withdrawAmount, nil, exitcode.Ok)
```

Comment:
Use adt.EmptyValue for the third parameter. An in-flight change will enforce this.

Reply:
@anorth should all `Send` calls with `nil` params in the market actor be changed to `adt.EmptyValue{}`?

Reply:
Yes please